### ISSUE TITLE
Lock-on for combat

### ISSUE THREAD

**Issue by metalgearsloth** [space-wizards/space-station-14#25637]
This is 50% ideaguys but combat is significantly harder in ss14 than ss13 due to pixel hunting; in ss13 mobs can't occupy same tiles in most circumstances which is partially the issue but mobs in ss13 also move predictably. One way to alleviate this is giving an action that continues targeting a specific person until toggled off. To make it not completely auto-aim you could add a threshold to where it goes back to manual aiming.

Ideally we would try to implement mob pushing first before experimenting with lock-on.

**Comment by PJB3005** [comment]
I mean this is what wide swings aim to solve in the first place, isn't it?

**Comment by metalgearsloth** [comment]
Melee's not too bad but non-predicted gun combat makes shooting close quarters largely ping based.

**Comment by metalgearsloth** [comment]
Right-click lock-on left-click swing could also solve it maybe, where no-lock is wide-swing and lock is click attack, though this is still months away before I'd try testing it.

**Comment by Errant-4** [comment]
Some sort of lock-on could also mechanically facilitate "I am threatening you with my gun" situations, if it was telegraphed. 
And more relevantly, it could mean that not immediately shooting/stunning a suspected target will not completely lose you the initiative: if they suddenly attack you while you are trying to talk to them or Examine them, you could already have them targeted

**Comment by Kadeo64** [comment]
Maybe it would be best to make it middle mouse click?

**Comment by Kadeo64** [comment]
lock-on could also be given to like, laser sights, if weapon attachment systems are ever considered

**Comment by Krunklehorn** [comment]
There's a line or two in gun code that adjusts the trajectory of bullets based on your strafe direction.

That either needs to go or lock-on needs to compensate for it.

**Comment by Krunklehorn** [comment]
> > There's a line or two in gun code that adjusts the trajectory of bullets based on your strafe direction.
> 
> Where?

SharedGunSystem.cs

```
var targetMapVelocity = gunVelocity + direction.Normalized() * speed;
```

**Comment by Krunklehorn** [comment]
https://github.com/space-wizards/space-station-14/assets/42424291/189771f7-7340-414c-aaf2-1fa00018df3c

If you aim where you intend the bullet to land, you actually miss 100% of the time.

**Comment by QuietlyWhisper** [comment]
> If you aim where you intend the bullet to land, you actually miss 100% of the time.

Oh no wonder guns feel jank at times...

**Comment by Krunklehorn** [comment]
Whether or not it makes sense from a realism vs. gameplay perspective is it's own thread.

Regardless, a lock-on mechanic would be useless if it doesn't address momentum somehow.

**Comment by Tayrtahn** [comment]
> To me this looks like a take on Momentum Conservation Principle, which in this simple case means all projectiles inherit parent velocity prior to being propelled in the target direction.

That's great for slow-moving projectiles, but the issue is that bullets only move as slowly as they do for gameplay reasons. If they moved at more realistic speeds, the effect of inheriting the parent's velocity would be negligible.

If you throw a baseball out a car window it definitely should land quite a bit farther forward than you aimed; but if you shoot a gun out a car window, it's not going to miss the target by several feet. The fact that SS14's bullets move at speeds more like a baseball is a gameplay concession. They probably should aim like hitscan weapons, even if the projectile movement isn't that fast.

Not sure what the right solution is here. Maybe a flag on projectiles that determines if they should inherit velocity? Feels weird disabling physical realism for gameplay, but the slow bullets are already a bit of an oddity there.

**Comment by K-Dynamic** [comment]
> > If you aim where you intend the bullet to land, you actually miss 100% of the time.
> 
> Oh no wonder guns feel jank at times...

And I thought it was just high ping

**Comment by Ilya246** [comment]
making projectiles not inherit shooter velocity will make space combat worse

**Comment by Krunklehorn** [comment]
> making projectiles not inherit shooter velocity will make space combat worse

By space combat are you referring to spaceships or people in jetpacks?

**Comment by Ilya246** [comment]
> > making projectiles not inherit shooter velocity will make space combat worse
> 
> By space combat are you referring to spaceships shooting other spaceships or people shooting each other in jetpacks?

both

**Comment by K-Dynamic** [comment]
Tbh I kinda agree with lordeclipse, not sure if either option would be better

I'd also examine ttk on guns but that would be off-topic and changed with surgery/woundmed

**Comment by Kadeo64** [comment]
we should get dodge rolling with lock-on so we can play (insert any game with dodge rolling and lock on here, seriously there's like 30)

**Comment by Tayrtahn** [comment]
This means that technically we can "curve" bullets like in that movie, huh?

Good point about space combat. Tricky puzzle.

**Comment by Jezithyr** [comment]
Lock on would also solve some issues with targeting 😏 and the way I'd like to do targeting requires players to be able to "select a target". Locking on would be a good way to do that.

**Comment by Jezithyr** [comment]
*also* locking onto targets would allow us to implement different bullet scatter/hit chances for weapons depending on if you are locked on to a target or not to simulate "hipfire" vs aimed-fire. Not to mention is a nice QOL and accessibility feature that decreases pixelhunting. Hell, it'll make SS14 combat more playable on a gamepad which would be cool (Which helps disabled folks because most accessibility input devices map to gamepads NOT keyboards/mice)

**Comment by K-Dynamic** [comment]
> _also_ locking onto targets would allow us to implement different bullet scatter/hit chances for weapons depending on if you are locked on to a target or not to simulate "hipfire" vs aimed-fire. 

VATS but SS14 :trollface:
